google-cloud-storage/src/main/java/com/google/cloud/storage/PostPolicyV4.java

```diff
@@ -0,0 +1,378 @@
+                        conditions.add(match);
+                        break;
+                    case STARTS_WITH:
+                        JsonArray startsWith = new JsonArray();
+                        startsWith.add("starts-with");
+                        startsWith.add("$" + condition.element);
+                        startsWith.add(condition.value);
+                        conditions.add(startsWith);
+                        break;
+                    case CONTENT_LENGTH_RANGE:
+                        JsonArray contentLengthRange = new JsonArray();
+                        contentLengthRange.add("content-length-range");
+                        contentLengthRange.add(Integer.parseInt(condition.element));
+                        contentLengthRange.add(Integer.parseInt(condition.value));
+                        conditions.add(contentLengthRange);
+                        break;
+                }
+            }
+            object.add("conditions", conditions);
+            object.addProperty("expiration", expiration);
+
+            String json = object.toString();
+            StringBuilder escapedJson = new StringBuilder();
+
+            //Certain characters in a policy must be escaped
+            for (char c : json.toCharArray()) {
+                if (c >= 128) { //is a unicode character
+                    escapedJson.append(String.format("\\u%04x", (int) c));
+                } else {
+                    switch (c) {
+                        case '\\': escapedJson.append("\\\\"); break;
+                        case '\b' : escapedJson.append("\\b"); break;
+                        case '\f' : escapedJson.append("\\f"); break;
+                        case '\n' : escapedJson.append("\\n"); break;
+                        case '\r' : escapedJson.append("\\r"); break;
+                        case '\t' : escapedJson.append("\\t"); break;
+                        case '\u000b' : escapedJson.append("\\v"); break;
+                        default : escapedJson.append(c);
+                    }
+                }
+            }
+            return escapedJson.toString();
+        }
+    }
+
+    enum ConditionV4Type {
+        MATCHES,
+        STARTS_WITH,
+        CONTENT_LENGTH_RANGE
+    }
+
+    /**
+     * Class for a specific POST policy document condition.
+     *
+     * @see <a href="https://cloud.google.com/storage/docs/authentication/signatures#policy-document"> Policy document</a>
+     */
+    public static final class ConditionV4 {
+        ConditionV4Type type;
+        String element;
+        String value;
+
+        private ConditionV4(ConditionV4Type type, String element, String value) {
+            this.type = type;
+            this.element = element;
+            this.value = value;
+        }
+
+        public static ConditionV4 newV4Condition(ConditionV4Type type, String element, String value) {
+            return new ConditionV4(type, element, value);
```

Comment:
Given PostConditionV4 requires users to use a related method. Should this class have a private static method helper as well? Also not sure how this is different from `private ConditionV4()`?